app/scripts/modules/core/src/pipeline/config/graph/PipelineGraph.tsx

```diff
@@ -1,11 +1,464 @@
+  onNodeClick: (node: IPipelineGraphNode) => void;
   pipeline?: IPipeline;
-  execution: IExecution;
   viewState: IExecutionViewState;
-  onNodeClick: (node: IPipelineNode) => void;
-  shouldValidate?: boolean;
+}
+
+export interface IPipelineGraphState {
+  allNodes: IPipelineGraphNode[];
+  graphHeight: number;
+  graphWidth: string;
+  labelOffsetX: number;
+  labelOffsetY: number;
+  maxLabelWidth: number;
+  nodeRadius: number;
+  phaseCount: number;
+  rowHeights: number[];
+}
+
+@autoBindMethods
+export class PipelineGraph extends React.Component<IPipelineGraphProps, IPipelineGraphState> {
+  private defaultNodeRadius = 8;
+  private defaultState: IPipelineGraphState = {
+    allNodes: [],
+    graphHeight: 0,
+    graphWidth: '',
+    labelOffsetX: this.defaultNodeRadius + 3,
+    labelOffsetY: this.defaultNodeRadius + 10,
+    maxLabelWidth: 0,
+    nodeRadius: this.defaultNodeRadius,
+    phaseCount: 0,
+    rowHeights: [],
+  };
+  private element: JQuery;
+  private graphStatusHash: string;
+  private graphVerticalPadding = 11;
+  private minExecutionGraphHeight = 40;
+  private minLabelWidth = 100;
+  private pipelineValidations: IPipelineValidationResults = { pipeline: [], stages: [] };
+  private rowPadding = 20;
+  private validationSubscription: Subscription;
+  private windowResize = this.handleWindowResize.bind(this);
+
+  constructor(props: IPipelineGraphProps) {
+    super(props);
+    this.state = this.defaultState;
+
+    // HACK: This is needed to update the node states in the graph based on the stage states.
+    //       Once the execution itself changes based on stage status, this can be removed.
+    if (this.props.execution) { this.graphStatusHash = this.props.execution.graphStatusHash; }
+  }
+
+  private highlight(node: IPipelineGraphNode, highlight: boolean): void {
+    if (node.isActive) { return; }
+    node.isHighlighted = highlight;
+    node.parentLinks.forEach((link: IPipelineGraphLink) => link.isHighlighted = highlight);
+    node.childLinks.forEach((link: IPipelineGraphLink) => link.isHighlighted = highlight);
+    this.applyAllNodes(this.state);
+  }
+
+  @Throttle(300)
+  private handleWindowResize(): void {
+    this.updateGraph(this.props);
+  }
+
+  private handleWheel(e: React.WheelEvent<HTMLDivElement>) {
+    // track and save the graph scroll position for executions so it doesn't get reset to
+    // zero every second due to repaint.
+    if (this.props.execution) {
+      PipelineGraphService.xScrollOffset[this.props.execution.id] = e.deltaX;
+    }
+  }
+
+  /**
+   * Used to draw inverse bezier curve between stages
+   */
+  private curvedLink(d: any) {
+    const sourceX = d.source.x + this.state.nodeRadius;
+    const targetX = d.target.x - this.state.nodeRadius;
+    return 'M' + sourceX + ',' + d.source.y
+        + 'C' + (sourceX + targetX) / 2 + ',' + d.source.y
+        + ' ' + (sourceX + targetX) / 2 + ',' + d.target.y
+        + ' ' + targetX + ',' + d.target.y;
```

Comment:
Is this easier to read or harder to read as a string template?
```javascript
const curve = (sourceX + targetX) / 2;
return `M${sourceX},${d.source.y}C${curve},${d.source.y} ${curve},${d.target.y} ${targetX},${d.target.y}`;
```

Reply:
Easier. I'll change it.